### DISCUSSION TITLE
Graphql Filter using the “where” method to check if a record contain empty array

### DISCUSSION THREAD

**Discussion by asdutoit** [strapi/strapi#6480] [Legacy Topics]
Hi Everyone, 

Does anyone know how to use the graphql “where” filter method to check if an array contains any items? Currently, I use the following query:

```
query devices {
  onts(where:{images_null: false}){
    id
    images{
      id
    }
  }
}
```

which returns:

```
{
  "data": {
    "onts": [
      {
        "id": "5ebfffa957c09c029fa7831c",
        "images": [
          {
            "id": "5ed3a5040889c464b4e5e07f"
          }
        ]
      },
      {
        "id": "5eccc3bb7c3b9d59351593de",
        "images": []
      },
      {
        "id": "5ece135e7c3b9d59351593df",
        "images": []
      }
    ]
  }
}
```

As you can see, 2x of the returned records contain empty arrays.
I would like the query to only return records that contain at least one item in the array.

Using Strapi v 3.0.1

Appreciate any assistance.

**Comment by hashinteractive** [comment]
@asdutoit did you ever happen to figure this out?

**Reply by asdutoit** [reply]
Hi @hashinteractive ,  No unfortunately not.
I got sidetracked and havent yet resumed my focus on this issue.  I assume you have the same query?

**Comment by iicdii** [comment]
What about
```js
query devices {
  onts(where: {images_ne: []}){
    id
    images {
      id
    }
  }
}
```
?

**Reply by hashinteractive** [reply]
Unfortunately this did not work for me either

**Comment by asdutoit** [comment]
Thanks, @iicdii, 

Unfortunately, it did not work.  
Ironically my original query actually now does work:

```
onts(where:{images_null: false}){
    id
    images{
      id
    }
  }
```

I made a change in the ONTs content-type by adding an images field directly to the ONT content-type instead of building a separate related IMAGES content-type.   This seemed to have done the trick.   However, I do suspect the issue in my original implementation stemmed from me having built the relationship between the two content-types wrong. 

1.  My graphql query:

```
query ONTS{
  onts(where:{images_null: false}){
    id
    serial_number
    images{
      id
      name
      url
    }
  }
}
```

2.  My ONTS content-type:

<img width="969" alt="Screenshot 2020-07-14 11 18 57" src="https://user-images.githubusercontent.com/20240370/87408494-d4efc780-c5c3-11ea-9bd8-3d4066581884.png">

**Comment by hashinteractive** [comment]
@asdutoit, interesting because I have a relationship (not a media relationship) and the graphqhl query that you implemented did not work for me either.

```
query ALL_CATEGORIES {
  categories(
    where: {
      subcategories_null: false
    }){
      name
      id
      subcategories {
        name
        id
      }
    }
}
```

I still get all of the categories returned even if the subcategories are empty (return an empty array).

The `subcategories_null: false` shouldn't work because it's not returning a null value and an empty array wouldn't evaluate to `null` and I also tried `subcategories_nin: []` as well to no avail.

**Reply by asdutoit** [reply]
Can you perhaps add screenshots of your content-types?

**Comment by hashinteractive** [comment]
I can confirm this works with media fields:

```
query ONTS{
  onts(where:{images_null: false}){
    id
    serial_number
    images{
      id
      name
      url
    }
  }
}
```

But it **does not** work on other types of relationships. 

Ie. If you have a content type called categories with a relationship to a products content type:
![image](https://user-images.githubusercontent.com/12477388/92113969-bb921d00-edb5-11ea-942d-e2282c2909c8.png)

You cannot filter Categories that do not have products (categories with no products) or in which the product array is empty:
ie. **non of these implementations work**

`http://localhost:1337/categories?products_null=true`
or
```
query ALL_CATEGORIES {
  categories(
    where: {
      products_null: true
    }){
      name
      id
      subcategories {
        name
        id
      }
    }
}
```

Nor does this work:
`http://localhost:1337/categories?products_in=[]`
or
```
query ALL_CATEGORIES {
  categories(
    where: {
      products_in: []
    }){
      name
      id
      subcategories {
        name
        id
      }
    }
}
```

Nor does this work:
`http://localhost:1337/categories?products_eq=[]`
or
```
query ALL_CATEGORIES {
  categories(
    where: {
      products: []
    }){
      name
      id
      subcategories {
        name
        id
      }
    }
}
```

Nor does this work:
`http://localhost:1337/categories?products.id_null=true`

**Reply by asdutoit** [reply]
Ah I see.   Yeah, I experience the same issue now upon testing.   
I see in later comments @derrickmehaffy mentioned they are aware of the issue and planning on addressing it. 
They are rockstars

**Reply by jerodfritz** [reply]
I have the same use case and ultimately a very similar issue.  We needed a way to create category hierarchies and menu structures so we made child/parent relationships on the collections.  To eb able to query for the top level items and vice versa we created an additional boolean field on the collection "primary" and just toggled it as true for the top level items to get around these types of queries.

**Comment by derrickmehaffy** [comment]
Currently it's not possible to filter on polymorphic relations (media/images in this case) it's something we are looking into with our database rework for Q4.

**Reply by hashinteractive** [reply]
@derrickmehaffy interestingly enough filtering on null (not null) images does work but other polymorphic relationships does not. I'm assuming because the relationship to media (upload_files) is a "has many" vs "belongs to many" which is possibly why it works on media upload filtering and not other polymorphic (belongs to many) relationships?

**Reply by derrickmehaffy** [reply]
Quite possible, the deep filtering and filtering on relations is a pain point we are well aware of. The fixes to them however are not simple hence why we are planning a massive rework of our database layer for Q3/Q4. 

https://portal.productboard.com/strapi/1-roadmap/c/73-database-layer-v4

**Comment by lucap84** [comment]
@asdutoit If you have a mandatory String field in images, you could try this workaround to obtain all the `onts` without relationships with `image`:

```query devices {
  onts(where:{images: { stringField_ncontains: "" } }) {
    id
    images{
      id
      stringField
    }
  }
}
